sup3r/utilities/loss_metrics.py

```diff
@@ -117,6 +117,103 @@ def __call__(self, x1, x2, sigma=1.0):
+            (n_observations, spatial_1, spatial_2, temporal)
+        axis : int
+            Axis to take derivative over
+        """
+        if axis == 1:
+            return tf.concat([x[:, 1:2] - x[:, 0:1],
+                              (x[:, 2:] - x[:, :-2]) / 2,
+                              x[:, -1:] - x[:, -2:-1]], axis=1)
+        elif axis == 2:
+            return tf.concat([x[..., 1:2, :] - x[..., 0:1, :],
+                              (x[..., 2:, :] - x[..., :-2, :]) / 2,
+                              x[..., -1:, :] - x[..., -2:-1, :]], axis=2)
+        elif axis == 3:
+            return tf.concat([x[..., 1:2] - x[..., 0:1],
+                              (x[..., 2:] - x[..., :-2]) / 2,
+                              x[..., -1:] - x[..., -2:-1]], axis=3)
+
+        else:
+            msg = (f'{self.__class__}._derivative received axis={axis}. '
+                   'This is meant to compute only temporal (axis=3) or '
+                   'spatial (axis=1/2) derivatives.')
+            raise ValueError(msg)
+
+    def _compute_md(self, x, fidx):
+        """Compute material derivative the feature given by the index fidx.
+        It is assumed that for a given feature index fidx there is a pair of
+        wind components u/v given by 2 * (fidx // 2) and 2 * (fidx // 2) + 1
+
+        Parameters
+        ----------
+        x : tf.tensor
+            synthetic output or high resolution data
+            (n_observations, spatial_1, spatial_2, temporal, features)
+        fidx : int
+            Feature index to compute material derivative for.
+        """
+        uidx = 2 * (fidx // 2)
+        vidx = 2 * (fidx // 2) + 1
+        # df/dt
+        x_div = self._derivative(x[..., fidx], axis=3)
+        # u * df/dx
+        x_div += tf.math.multiply(
+            x[..., uidx], self._derivative(x[..., fidx], axis=1))
+        # v * df/dy
+        x_div += tf.math.multiply(
+            x[..., vidx], self._derivative(x[..., fidx], axis=2))
+
+        return x_div
+
+    def __call__(self, x1, x2):
+        """Custom content loss that encourages accuracy of the material
+        derivative. This assumes that the first 2 * N features are N u/v
+        wind components at different hub heights and that the total number of
+        features is either 2 * N or 2 * N + 1
+
+        Parameters
+        ----------
+        x1 : tf.tensor
+            synthetic generator output
+            (n_observations, spatial_1, spatial_2, temporal, features)
+        x2 : tf.tensor
+            high resolution data
+            (n_observations, spatial_1, spatial_2, temporal, features)
+
+        Returns
+        -------
+        tf.tensor
+            0D tensor with loss value
+        """
+        hub_heights = x1.shape[-1] // 2
+
+        x1_div = tf.stack(
+            [self._compute_md(x1, fidx=i) for i in range(2 * hub_heights)])
+        x2_div = tf.stack(
+            [self._compute_md(x2, fidx=i) for i in range(2 * hub_heights)])
```

Comment:
Shouldn't this be `range(0, 2*hh, 2)`? Right now you're iterating from 0 to 2*hh and when i=1 I think that `self._compute_md()` will use i=(1,2) as u10m/v10m but i=(1,2) is really v10m/u100m

Actually i don't think thats quite right because `uidx = 2 * (fidx // 2)` but still, won't you be double calculating if you do `i in range(2*hh)`?